### PR TITLE
Pass Quantization parameters for quant nodes

### PR DESCRIPTION
**Pull request by nishantpdce**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19910 Change the quant-dequant node pattern for jit op substitution pass&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15120146/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19402 Pass Quantization parameters for quant nodes**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14995230/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19232 Add observer nodes for input data nodes&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14885485/)

This pass propagate the qparams calculated after calibration to the
quant nodes which will be used later for quantization

Differential Revision: [D14995230](https://our.internmc.facebook.com/intern/diff/D14995230/)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -1236,124 +1236,164 @@ def observer(self, x, name):
...
-            def forward(self, x):
-                x = F.relu(self.conv1(x))
-                return x
+    def test_insert_quantdequant(self):
+        # Test Class to generate default qparams
+        script_only = True
+        x1 = torch.ones([1, 1, 1, 1])
 
-        trace = testModule()
+        class TestGenQParam:
```
Comment (ZolotukhinM) [inline test/test_jit.py]: This test is huge, and it's very hard to understand what it checks. Please replace it with a small unit test that checks **only** what the pass is supposed to do.

File: test/test_jit.py [hunk 2]
```diff
@@ -1236,124 +1236,164 @@ def observer(self, x, name):
...
+            # happens as part of this jit pass
+            torch._C._jit_pass_constant_propagation(trace.graph)
+
+            test_helper_run_quantdequant(trace, x1)
+
+            # We expect to see quant-dequant node before and after
+            # both conv and relu nodes and at external output since relu
+            # is last node. Constant nodes correspond to params for the
+            # quantization nodes
+            FileCheck().check("quantize_linear").check_next("dequantize") \
```
Comment (ZolotukhinM) [inline test/test_jit.py]: Please use a new style of writing tests, an example can be found in `test_recursive_cse` test, a documentation is available here: https://github.com/pytorch/pytorch/blob/master/test/HowToWriteTestsUsingFileCheck.md. It allows to make the intention of FileCheck checks clearer.

File: torch/csrc/jit/passes/quantization.cpp [hunk 3]
```diff
@@ -39,40 +39,42 @@ int groups, bool benchmark, bool deterministic, bool cudnn_enabled) -> Tensor"
   return quantnodeLookup.find(n);
 }
 
-void insertQuantNodeParams(Node* quant, std::tuple<float, int> qparam) {
+void insertQuantNodeParams(Node* quant, float scale, int zero_point) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: This seems like a refactoring that doesn't change what the function is doing - please split such things into separate PRs. It makes reviewing and potential future debugging easier.

File: torch/csrc/jit/passes/quantization.cpp [hunk 4]
```diff
@@ -39,40 +39,42 @@ int groups, bool benchmark, bool deterministic, bool cudnn_enabled) -> Tensor"
...
       n->owningGraph()->create(at::Symbol::fromQualString(
         "aten::quantize_linear"));
   AT_ASSERTM(quant != nullptr, "Failed to create quant node");
   quant->output()->setUniqueName(v->uniqueName() + ".quant");
+  quant->setScope(n->scope());
+  return quant;
+}
 
+// Create Dequant node
+Node* createDeQuantNode(Value* v, Node* n) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Same here.

File: torch/csrc/jit/passes/quantization.cpp [hunk 5]
```diff
@@ -81,18 +83,19 @@ void addQuantDeQuantNodes(Value* v) {
...
+
+  insertQuantNodeParams(quant, std::get<1>(qparam),
+    std::get<2>(qparam));
   dequant->addInput(quant->output());
 }
 
 // Insert Quant-Dequant node pair for specific input to node n
-void addQuantDeQuantNodesForInput(Value* v, Node* n) {
+void addQuantDeQuantNodesForInput(Value* v, Node* n,
+  std::tuple<std::string, float, int>& qparam) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: You just moved away from using tuple in the function above - maybe flatten this one as well then?

File: torch/csrc/jit/passes/quantization.cpp [hunk 6]
```diff
@@ -104,10 +107,29 @@ void addQuantDeQuantNodesForInput(Value* v, Node* n) {
...
 
+template <typename... ArgT>
+bool matchQParamDictKeytoObserver(Node* n,
+  std::unordered_map<std::string, std::tuple<ArgT...>>& qparam_dict,
+  std::unordered_map<Value*, std::tuple<ArgT...>>& qparam_value_dict) {
+  if (n->kind() != prim::PythonOp) {
+    return false;
+  }
+  Value* vname = n->inputs()[1];
+  IValue valuekey = toIValue(vname).value();
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Maybe add `AT_ASSERT(vname->has_value())` before that?

File: torch/csrc/jit/passes/quantization.cpp [hunk 7]
```diff
@@ -104,10 +107,29 @@ void addQuantDeQuantNodesForInput(Value* v, Node* n) {
...
 }
 
+template <typename... ArgT>
+bool matchQParamDictKeytoObserver(Node* n,
+  std::unordered_map<std::string, std::tuple<ArgT...>>& qparam_dict,
+  std::unordered_map<Value*, std::tuple<ArgT...>>& qparam_value_dict) {
+  if (n->kind() != prim::PythonOp) {
+    return false;
+  }
+  Value* vname = n->inputs()[1];
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Check number of inputs before accessing 2nd element.

File: torch/csrc/jit/passes/quantization.cpp [hunk 8]
```diff
@@ -104,10 +107,29 @@ void addQuantDeQuantNodesForInput(Value* v, Node* n) {
...
+  insertQuantNodeParams(quant, std::get<1>(qparam),
+    std::get<2>(qparam));
   dequant->addInput(quant->output());
 }
 
+template <typename... ArgT>
+bool matchQParamDictKeytoObserver(Node* n,
+  std::unordered_map<std::string, std::tuple<ArgT...>>& qparam_dict,
+  std::unordered_map<Value*, std::tuple<ArgT...>>& qparam_value_dict) {
+  if (n->kind() != prim::PythonOp) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: A node might be `PythonOp`, but not an observer. In that case this will crash.

File: torch/csrc/jit/passes/quantization.cpp [hunk 9]
```diff
@@ -104,10 +107,29 @@ void addQuantDeQuantNodesForInput(Value* v, Node* n) {
   // Attach inputs to quant and dequant nodes
   quant->addInput(v);
   // Default Quant Params <Scale:1.0, ZeroPoint:0>
-  insertQuantNodeParams(quant, std::make_tuple(1.0, 0));
+  insertQuantNodeParams(quant, std::get<1>(qparam),
+    std::get<2>(qparam));
   dequant->addInput(quant->output());
 }
 
+template <typename... ArgT>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Why do we need to templatize it when we know exactly what types we expect and they are always the same?

File: torch/csrc/jit/passes/quantization.cpp [hunk 10]
```diff
@@ -223,7 +245,9 @@ void InsertObserverNodes(script::Method* method,
   }
 }
 
-void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph) {
+void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph,
+  std::unordered_map<std::string, std::tuple<std::string,
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: `qparam_dict` should be `const` I suppose.

File: torch/csrc/jit/passes/quantization.cpp [hunk 11]
```diff
@@ -235,6 +259,14 @@ void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph) {
   std::vector<Value*> quantOutputs;
   std::unordered_set<Value*> valueLookup;
 
+  // Observer nodes to remove from graph
+  std::vector<Node*> nodes_to_remove;
+
+  // Create value:qparam dict which guarantees no
+  // name conflict between passes
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: I don't understand this comment, could you elaborate please?

File: torch/csrc/jit/passes/quantization.h [hunk 12]
```diff
@@ -40,9 +40,10 @@ TORCH_API void InsertObserverNodes(
  * quantizatable tensors - later passes only cleanup the IR and
  * make sure the model runs faster/consumes less memory.
  *
- * TODO: This should also take a qparam-map as an input.
  */
-TORCH_API void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph);
+TORCH_API void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph,
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.h]: Please don't forget to update the comment.

File: torch/csrc/jit/passes/quantization.cpp [hunk 13]
```diff
@@ -245,6 +277,15 @@ void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph) {
         blocks_to_visit.push(subblock);
       }
 
+      if (matchQParamDictKeytoObserver<std::string, float, int>
+        (n, qparam_dict, qparam_value_dict)) {
+        // This is the observer node. Mark it and the second input
+        // constant nod for deletion.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: typo: nod

**Review by ZolotukhinM — CHANGES_REQUESTED** [review]
Please see inline!

File: torch/csrc/jit/passes/quantization.cpp (hunk 8 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: yeah thats why  we match the second argument as confirmation it is indeed observer. Let me add more checks to ensure it does not crash

File: torch/csrc/jit/passes/quantization.cpp (hunk 9 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: no strong reason, its just this api is matching against key only, if in future tuple changes, this api can remain same. if it is major concern i can remove also

File: test/test_jit.py [hunk 14]
```diff
@@ -269,6 +269,28 @@ def _sum_of_list(tensorlist):
         s += t.sum()
     return s
 
+# helper function to generate test qparam
+
+
+def _helper_generate_qparam(trace, x1):
```
Comment (raghuramank100) [inline test/test_jit.py]: Minor nit: Can we call the input as 'input_data' not x1?. Also, can we call the input model, my_script_model instead of trace as we arent tracing it?

File: test/test_jit.py [hunk 15]
```diff
@@ -1232,7 +1256,9 @@ def forward(self, x):
         # to insert quant-dequant nodes for quantizable tensors. The type analysis
         # happens as part of this jit pass
         torch._C._jit_pass_constant_propagation(trace.graph)
-        self.run_pass('insert_quantdequant', trace.graph)
+
```
Comment (raghuramank100) [inline test/test_jit.py]: Lets call this my_script_module instead of trace

File: test/test_jit.py [hunk 16]
```diff
@@ -269,6 +269,28 @@ def _sum_of_list(tensorlist):
...
+            if name not in self.qparam_dict:
+                self.qparam_dict[name] = []
+            self.qparam_dict.update({name : (self.qscheme, 1.0, 0)})
+            return x
+    activationObj = TestGenQParam(qscheme='per_tensor_quant')
+    torch._C._jit_pass_insert_observers(trace, activationObj.observer)
+    trace.forward(x1)
+    # qparam_dict
+    qparam_dict = {}
+    qparam_dict.update(activationObj.qparam_dict)
```
Comment (raghuramank100) [inline test/test_jit.py]: Cant we just return activationObj.qparam_dict here?

File: torch/csrc/jit/passes/quantization.cpp [hunk 17]
```diff
@@ -40,39 +40,45 @@ int groups, bool benchmark, bool deterministic, bool cudnn_enabled) -> Tensor"};
...
+  Value* zeropoint_v = quant->owningGraph()->insertConstant(zero_point);
+  quant->addInput(scale_v);
+  quant->addInput(zeropoint_v);
 }
 
-// Create Quant-Dequant node pair for quantizable Value
-std::pair<Node*, Node*> createQuantDeQuantNodes(Value* v, Node* n) {
-  Node* quant = n->owningGraph()->create(
-      at::Symbol::fromQualString("aten::quantize_linear"));
+// Create Quant Node
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: This is not a very helpful comment, it just duplicates what the function name says. At the same time, I don't know what is V and N that we pass into here.

File: torch/csrc/jit/passes/quantization.cpp [hunk 18]
```diff
@@ -40,39 +40,45 @@ int groups, bool benchmark, bool deterministic, bool cudnn_enabled) -> Tensor"};
   return quantnodeLookup.find(n);
 }
 
-void insertQuantNodeParams(Node* quant, std::tuple<float, int> qparam) {
+void insertQuantNodeParams(Node* quant,
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Please run clang-format on your PR (and install clang-format hooks so that you don't have to do it manually every time).

File: torch/csrc/jit/passes/quantization.cpp [hunk 19]
```diff
@@ -40,39 +40,45 @@ int groups, bool benchmark, bool deterministic, bool cudnn_enabled) -> Tensor"};
...
+  std::tuple<std::string, float, int>& qparam) {
+  auto& scale = std::get<1>(qparam);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Why can't we do it inside `createQuantNode`?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see inline comments.Thanks!

File: torch/csrc/jit/passes/quantization.cpp [hunk 20]
```diff
@@ -81,18 +87,17 @@ void addQuantDeQuantNodes(Value* v) {
 
   // Attach inputs to quant and dequant nodes
   quant->addInput(v);
-  // Default Quant Params <Scale:1.0, ZeroPoint:0>
-  insertQuantNodeParams(quant, std::make_tuple(1.0, 0));
+  insertQuantNodeParams(quant, qparam);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Do we need to pass the same parameters to the `dequant` node as well?

File: torch/csrc/jit/passes/quantization.cpp [hunk 21]
```diff
@@ -81,18 +87,17 @@ void addQuantDeQuantNodes(Value* v) {
...
   dequant->addInput(quant->output());
 }
 
 // Insert Quant-Dequant node pair for specific input to node n
-void addQuantDeQuantNodesForInput(Value* v, Node* n) {
+void addQuantDeQuantNodesForInput(Value* v, Node* n,
+  std::tuple<std::string, float, int>& qparam) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: `const std::tuple&` ?

File: torch/csrc/jit/passes/quantization.cpp [hunk 22]
```diff
@@ -103,11 +108,31 @@ void addQuantDeQuantNodesForInput(Value* v, Node* n) {
...
+  AT_ASSERT(toIValue(vname).has_value());
+  IValue valuekey = toIValue(vname).value();
+  Value* v = n->inputs()[0];
+  auto it = qparam_dict.find(valuekey.toStringRef());
+  if (it == qparam_dict.end()) {
+    return false;
+  }
+  // This is observer node for Value v
+  qparam_value_dict.emplace(v, it->second);
+  return true;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: What if my model have a PythonOp with 2 operands, and it is not an observer?

File: torch/csrc/jit/passes/quantization.cpp [hunk 23]
```diff
@@ -203,7 +228,9 @@ void InsertObserverNodes(const std::unique_ptr<script::Method>& method,
   }
 }
 
-void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph) {
+void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph,
+  const std::unordered_map<std::string, std::tuple<std::string,
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Maybe create a struct instead of the tuple? It's hard to tell what fields is what.

File: torch/csrc/jit/passes/quantization.cpp [hunk 24]
```diff
@@ -215,6 +242,15 @@ void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph) {
   std::vector<Value*> quantOutputs;
   std::unordered_set<Value*> valueLookup;
 
+  // Observer nodes to remove from graph
+  std::vector<Node*> nodes_to_remove;
+
+  // Create value:qparam dict. Even if ssa-name changed between
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: While I understand what you're trying to say here, I don't think that message reflects what's going on in the code. The dictionary we create here has nothing to do with guarantees about correct mapping, it's just a convenience thing at this point. All guarantees were provided at the time of construction of observer nodes and the input dictionary. Could you think of a better wording here please?

File: torch/csrc/jit/passes/quantization.cpp [hunk 25]
```diff
@@ -225,6 +261,15 @@ void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph) {
         blocks_to_visit.push(subblock);
       }
 
+      if (matchQParamDictKeytoObserver<std::string, float, int>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: This function does two separate things, which makes it more confusing than it could be:
1) It checks if the node is observer
2) If the node is observer, it finds and extracts the corresponding value from `qparam_dict`.
Could you please split it?

File: test/test_jit.py [hunk 26]
```diff
@@ -1291,7 +1325,9 @@ def forward(self, x):
         # to insert quant-dequant nodes for quantizable tensors. The type analysis
         # happens as part of this jit pass
         torch._C._jit_pass_constant_propagation(trace.graph)
-        self.run_pass('insert_quantdequant', trace.graph)
+
+        qparam_dict = _helper_generate_qparam(trace, x1)
```
Comment (ZolotukhinM) [inline test/test_jit.py]: Instead of actually running the model with observers, could you just create the appropriate IR with a qparam_dict as-if it was produced by observer-pass? The test should ideally test **only** the pass in question, not all the stuff it might work with.

File: test/test_jit.py [hunk 27]
```diff
@@ -1216,10 +1238,12 @@ def observe(x, name):
         self.assertEqual(value_stats['z'][1], x2 - y2)
 
     def test_insert_quantdequant_consecutive_qnodes_script(self):
+        x1 = torch.ones([1, 1, 1, 1])
+
         class testModule(torch.jit.ScriptModule):
             def __init__(self):
                 super(testModule, self).__init__()
-                self.conv1 = nn.Conv2d(1, 20, 5, 1)
+                self.conv1 = nn.Conv2d(1, 1, 1, 1)
```
Comment (ZolotukhinM) [inline test/test_jit.py]: This seems an unrelated (good) change, could you please split it into a separate PR?

File: test/test_jit.py [hunk 28]
```diff
@@ -269,6 +269,28 @@ def _sum_of_list(tensorlist):
         s += t.sum()
     return s
 
+# helper function to generate test qparam
+
+
```
Comment (ZolotukhinM) [inline test/test_jit.py]: Wouldn't these two lines break flake8?

**Review by ZolotukhinM — CHANGES_REQUESTED** [review]
It's looking better, please find some comments inline!

PS: I find the commit message a bit confusing - it sounds as if you added a new pass that propagated qparams, while in fact you just fixed the existing insert-quant-dequant pass to actually use qparams.

**Review by jerryzh168 — APPROVED** [review]
lgtm
